File: cpp/Exercise/LC0033_SearchInRotatedSortedArray.cpp
#include "bits/stdc++.h"

using namespace std;

class Solution {
public:
    int search(vector<int> &nums, int target) {
        int left = 0, right = nums.size() - 1;

        // 移动左右指针
        while (left < right) {
            int mid = (left + right) / 2;
            // 异或操作：1 个为真或三个为真，结果为真
            if (nums[0] > target ^ nums[0] > nums[mid] ^ target > nums[mid]) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }

        return left == right && nums[left] == target ? left : -1;
    }
};

File: java/src/LeetCode/LC0033_SearchInRotatedSortedArray.java
package LeetCode;

public class LC0033_SearchInRotatedSortedArray {

    public static int search(int[] nums, int target) {

        if (nums.length < 1) {
            return -1;
        }

        int i = 0;
        int j = nums.length - 1;

        if (nums[i] > target && nums[j] < target) {
            return -1;
        }

        if (nums[i] == target) {
            return i;
        } else if (nums[j] == target) {
            return j;
        }

        while (i < j && nums[i] < target) {
            i++;
            if (nums[i] == target) {
                return i;
            }
        }

        while (i < j && nums[j] > target) {
            j--;
            if (nums[j] == target) {
                return j;
            }
        }

        return -1;

    }

}
